src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -236,151 +265,217 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                            statement_index: 0,
+                            block: *bb,
+                        })
+                        .filter(|s| visited_locations.insert(*s))
+                        .map(|s| {
+                            if self.is_back_edge(location, s) {
+                                match outmost_back_edge {
+                                    None => {
+                                        outmost_back_edge = Some(location);
+                                    }
+
+                                    Some(back_edge)
+                                        if location.dominates(back_edge, &self.dominators) =>
+                                    {
+                                        outmost_back_edge = Some(location);
+                                    }
+
+                                    Some(_) => {}
+                                }
+                            }
+
+                            s
+                        }),
+                );
             }
+        }
+
+        None
+    }
+
+    /// true if the `from` location can reach the any of `to_any` locations without going through
+    /// any backedges
+    fn can_reach_ignoring_backedges(&self, from: Location, to: Location) -> bool {
+        let mut visited_locations = FxHashSet::default();
+        let mut pending_locations = VecDeque::new();
+        visited_locations.insert(from);
+        pending_locations.push_back(from);
+        debug!("can_reach_ignoring_backedges: from={:?} to={:?}", from, to,);
+
+        while let Some(location) = pending_locations.pop_front() {
+            debug!(
+                "can_reach_ignoring_backedges: location={:?} pending_locations={:?}
+                   visited_locations={:?}",
+                location, pending_locations, visited_locations
+            );
+            if to == location {
+                // We've managed to reach the use location
+                debug!("can_reach_ignoring_backedges: found!");
+                return true;
+            }
+
+            let block = &self.mir.basic_blocks()[location.block];
 
-            // Keep track of where we have visited.
-            visited_locations.push(location);
+            if location.statement_index < block.statements.len() {
+                let successor = location.successor_within_block();
+                if visited_locations.insert(successor) {
+                    pending_locations.push_back(successor);
+                }
+            } else {
+                pending_locations.extend(
+                    block
+                        .terminator()
+                        .successors()
+                        .map(|bb| Location {
+                            statement_index: 0,
+                            block: *bb,
+                        })
+                        .filter(|s| visited_locations.insert(*s))
+                        .filter(|s| !self.is_back_edge(location, *s) || to == *s),
```

Comment:
I think that a usable definition of "a cfg node N is inside the loop whose header is the cfg node L" is:
1. L dominates N
2. there is a path from N to L that goes only through nodes that L dominates.

Because:
1. `L` dominates all the nodes within the loop, so that if `N` is within the loop and has a path within the loop that goes back to `L`, the condition will be true.
2. If `N` is outside of the loop but dominated by `L`, then going back to `L` will require going through an outer loop header, which will dominate `N` and therefore not be dominated by it.

Reply:
another class of "confusible" cases:
```Rust
loop {
    loop {
        if cond { break; }
        borrow(&mut v);
    }
    use(v);
}
```
is the same MIR as
```Rust
'b: loop {
    if cond {
        use(v);
    } else {
        borrow(&mut v);
    }
}
```

but the error message should be different

Reply:
So, given the second example, I think that any CFG-based model will be fairly conservative, so I'm ok with this